src/compiler/checker.ts

```diff
@@ -16633,7 +16633,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 if (getTypeParameterFromMappedType(mappedType) === getActualTypeVariable(type)) {
                     const typeParameter = getHomomorphicTypeVariable(mappedType);
                     if (typeParameter) {
-                        const constraint = getConstraintOfTypeParameter(typeParameter);
+                        const constraint = getConstraintOfTypeParameter(getConditionalFlowTypeOfType(typeParameter, parent));

```

Comment:
I will note that this call is at least a bit cursed; `getConditionalFlowTypeOfType` returns `Type`, not `TypeParameter`.

Reply:
yeee, this is really cursed 😬 since the result of this function was accepted as a valid type to `getConstraintOfTypeParameter`... I didn't really recheck what's happening here.

I'm changing this to a draft as I also noticed now that it needs extra consideration as to how it interacts with `getResolvedApparentTypeOfMappedType`

Reply:
I'm not yet sure what to do about it, especially in light of https://github.com/microsoft/TypeScript/pull/57801 that only treats mapped types ar arrays/tuples when every constituent of an intersection is array/tuple.

In this situation the constraint of the substitution type created by `getConditionalFlowTypeOfType` is smth like `Schema & SchemaArray` (so `(SchemaArray | SchemaObject | LiteralType) & SchemaArray`). The user's intention is clearly to treat the narrowed-down type as an array/tuple but this intersection type doesn't quite guarantee us that. Even though often an intersection like this is meant to "filter" it isn't exactly that.

I'll have to think about this one more but in the meantime I realized that there is a bug here that can be fixed separately: https://github.com/microsoft/TypeScript/pull/60084